python/ray/tune/cloud.py

```diff
@@ -0,0 +1,316 @@
+from ray.util import PublicAPI
+
+
+def is_cloud_target(target: str):
+    return any(target.startswith(prefix) for prefix in ALLOWED_REMOTE_PREFIXES)
+
+
+def _clear_bucket(bucket: str):
+    if not is_cloud_target(bucket):
+        raise ValueError(
+            f"Could not clear bucket contents: "
+            f"Bucket `{bucket}` is not a valid or supported cloud target.")
+
+    try:
+        if bucket.startswith(S3_PREFIX):
+            subprocess.check_call(
+                ["aws", "s3", "rm", "--recursive", "--quiet", bucket])
+        elif bucket.startswith(GS_PREFIX):
+            subprocess.check_call(["gsutil", "-m", "rm", "-f", "-r", bucket])
+        elif bucket.startswith(HDFS_PREFIX):
+            subprocess.check_call(["hdfs", "dfs", "-rm", "-r", bucket])
+
+    except Exception as e:
+        logger.warning(
+            f"Caught exception when clearing bucket `{bucket}`: {e}")
+
+
+def _download_from_bucket(bucket: str, local_path: str):
+    if not is_cloud_target(bucket):
+        raise ValueError(
+            f"Could not download from bucket: "
+            f"Bucket `{bucket}` is not a valid or supported cloud target.")
+
+    if bucket.startswith(S3_PREFIX):
+        subprocess.check_call(
+            ["aws", "s3", "cp", "--recursive", "--quiet", bucket, local_path])
+    elif bucket.startswith(GS_PREFIX):
+        subprocess.check_call(["gsutil", "-m", "cp", "-r", bucket, local_path])
+    elif bucket.startswith(HDFS_PREFIX):
+        subprocess.check_call(["hdfs", "dfs", "-get", bucket, local_path])
+
+
+def _upload_to_bucket(bucket: str, local_path: str):
+    if not is_cloud_target(bucket):
+        raise ValueError(
+            f"Could not download from bucket: "
+            f"Bucket `{bucket}` is not a valid or supported cloud target.")
+
+    if bucket.startswith(S3_PREFIX):
+        subprocess.check_call(
+            ["aws", "s3", "cp", "--recursive", "--quiet", local_path, bucket])
+    elif bucket.startswith(GS_PREFIX):
+        subprocess.check_call(["gsutil", "-m", "cp", "-r", local_path, bucket])
+    elif bucket.startswith(HDFS_PREFIX):
+        subprocess.check_call(["hdfs", "dfs", "-put", local_path, bucket])
+
+
+@PublicAPI(stability="beta")
+class TrialCheckpoint:
+    def __init__(self,
+                 local_path: Optional[str] = None,
+                 cloud_path: Optional[str] = None,
+                 metrics: Optional[Dict[str, Number]] = None):
+        self.local_path = local_path
+        self.cloud_path = cloud_path
+
+        self.metrics = metrics or {}
+
+    def __str__(self):
+        return self.local_path or self.cloud_path
+
+    def __repr__(self):
+        return (f"<TrialCheckpoint "
+                f"local_path={self.local_path}, "
+                f"cloud_path={self.cloud_path}"
+                f">")
+
+    def download(self,
+                 cloud_path: Optional[str] = None,
+                 local_path: Optional[str] = None,
+                 overwrite: bool = False,
```

Comment:
The overwrite option as well as the force download is a bit confusing to me as far as API is concerned.

Is that because a pair of local and cloud path won't cover it? Somehow we will change local path in the middle?

If possible, I would prefer to have a very scoped API and corresponding functionality.

Reply:
The reason why I'd like to keep `overwrite` is that I imagine a use case where a checkpoint to be served will be stored in the same local dir or cloud dir, e.g. something like `checkpoint.save("/home/ray/serve_checkpoint")`. If this location is updated with a newer checkpoint after training again, it is useful to be able to overwrite it from the API rather than having to delete it manually before.

Though after reviewing the code I'm not married to this API. I think the use case might be common enough to warrant the API, but happy to hear your and other people's thoughts on this

Reply:
Ah there is actually one CI test where we rely on this, but we could potentially change the scope.

Reply:
I guess overwrite is at least intuitive to understand, and seems useful.
update_local_path though, is a pretty confusing parameter.
why would we not want to update it? is it really necessary?

Reply:
We usually won't want to update it if it points to a temporary directory. Otherwise we can default to updating it whenever it is not set. I'll see if I can simplify this in the code

Reply:
I removed this from the API - thanks for the suggestion